Update GeneratorPythia8.cxx for Pythia 8.310+

cholmcc opened the pull request:
Pythia since version 8.310, no longer contains the header `Pythia8/HIUserHooks.h` but instead has `Pythia8/HIInfo.h`.  See also [here](https://pythia.org/history/).

cholmcc committed:
Update GeneratorPythia8.cxx for Pythia 8.310+

Pythia since version 8.310, no longer contains the header `Pythia8/HIUserHooks.h` but instead has `Pythia8/HIInfo.h`.  See also [here](https://pythia.org/history/).

alibuild commented:
Error while checking build/O2/fullCI for 51f8c90b9985de842294ddc04036a8cfb05d8f0a at 2024-03-11 20:53:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
task timeout reached .. killing all processes


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/6486a8022b90fac466fea7b5a028bc7a1cc158ef/slc8_x86-64/o2checkcode/1.0-local207/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12845/51f8c90b9985de842294ddc04036a8cfb05d8f0a/build_O2_fullCI/pretty.html).

cholmcc commented:
Error in `build/O2/fullCI` is a _false positive_ - some task timed out - again!

cholmcc committed:
More fixes for newer Pythia

In newer (>= 8.310) the `SubCollision` objects may not be stored with
the `HIInfo` object.  Thus, we cannot rely on those to be present when
calculating $N_{\mathrm{coll}}$, $N_{\mathrm{part}}$, and so on.

Instead, we must use the quantities stored directly in the `HIInfo`
object.

This PR fixes that issue by protecting on the Pythia version and falling
back to the information stored in `HIInfo`.

See also JIRA [issue](https://its.cern.ch/jira/browse/O2-4699).
